### Day 15 Regular Expressions in Python

### Recap

- Modules and Packages
    - Builtin Packages/Modules
        - math, random, os
    - User Defined Pakages/Modules

### Today's Objectives
- Regular Expressions(Regex)

In [1]:
s = """Python is an interpreted high-level general-purpose programming language. Python's design philosophy emphasizes code readability with its notable use of significant indentation. Wikipedia
Developer: Python Software Foundation
Stable release: 3.9.5 / 3 May 2021; 29 days ago
Preview release: 3.10.0b1 / 3 May 2021; 29 days ago
Typing discipline: Duck, dynamic, strong typing; gradual (since 3.5, but ignored in CPython)
First appeared: February 1991; 30 years ago
Paradigm: Multi-paradigm: object-oriented, procedural (imperative), functional, structured, reflective"""

In [2]:
print(list(s))

['P', 'y', 't', 'h', 'o', 'n', ' ', 'i', 's', ' ', 'a', 'n', ' ', 'i', 'n', 't', 'e', 'r', 'p', 'r', 'e', 't', 'e', 'd', ' ', 'h', 'i', 'g', 'h', '-', 'l', 'e', 'v', 'e', 'l', ' ', 'g', 'e', 'n', 'e', 'r', 'a', 'l', '-', 'p', 'u', 'r', 'p', 'o', 's', 'e', ' ', 'p', 'r', 'o', 'g', 'r', 'a', 'm', 'm', 'i', 'n', 'g', ' ', 'l', 'a', 'n', 'g', 'u', 'a', 'g', 'e', '.', ' ', 'P', 'y', 't', 'h', 'o', 'n', "'", 's', ' ', 'd', 'e', 's', 'i', 'g', 'n', ' ', 'p', 'h', 'i', 'l', 'o', 's', 'o', 'p', 'h', 'y', ' ', 'e', 'm', 'p', 'h', 'a', 's', 'i', 'z', 'e', 's', ' ', 'c', 'o', 'd', 'e', ' ', 'r', 'e', 'a', 'd', 'a', 'b', 'i', 'l', 'i', 't', 'y', ' ', 'w', 'i', 't', 'h', ' ', 'i', 't', 's', ' ', 'n', 'o', 't', 'a', 'b', 'l', 'e', ' ', 'u', 's', 'e', ' ', 'o', 'f', ' ', 's', 'i', 'g', 'n', 'i', 'f', 'i', 'c', 'a', 'n', 't', ' ', 'i', 'n', 'd', 'e', 'n', 't', 'a', 't', 'i', 'o', 'n', '.', ' ', 'W', 'i', 'k', 'i', 'p', 'e', 'd', 'i', 'a', '\n', 'D', 'e', 'v', 'e', 'l', 'o', 'p', 'e', 'r', ':', ' ', 'P'

In [4]:
li = []

for char in s:
    if char.isupper():
        li.append(char)
print(li)

['P', 'P', 'W', 'D', 'P', 'S', 'F', 'S', 'M', 'P', 'M', 'T', 'D', 'C', 'P', 'F', 'F', 'P', 'M']


In [5]:
li = []

for char in s:
    if not char.isalnum():    
        li.append(char)
print(li)

[' ', ' ', ' ', ' ', '-', ' ', '-', ' ', ' ', '.', ' ', "'", ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', '.', ' ', '\n', ':', ' ', ' ', ' ', '\n', ' ', ':', ' ', '.', '.', ' ', '/', ' ', ' ', ' ', ';', ' ', ' ', ' ', '\n', ' ', ':', ' ', '.', '.', ' ', '/', ' ', ' ', ' ', ';', ' ', ' ', ' ', '\n', ' ', ':', ' ', ',', ' ', ',', ' ', ' ', ';', ' ', ' ', '(', ' ', '.', ',', ' ', ' ', ' ', ' ', ')', '\n', ' ', ':', ' ', ' ', ';', ' ', ' ', ' ', '\n', ':', ' ', '-', ':', ' ', '-', ',', ' ', ' ', '(', ')', ',', ' ', ',', ' ', ',', ' ']


In [6]:
print(len(s))

565


- The loop will iterate for 565 times

#### Replace all special characters with **

In [7]:
print(s)

Python is an interpreted high-level general-purpose programming language. Python's design philosophy emphasizes code readability with its notable use of significant indentation. Wikipedia
Developer: Python Software Foundation
Stable release: 3.9.5 / 3 May 2021; 29 days ago
Preview release: 3.10.0b1 / 3 May 2021; 29 days ago
Typing discipline: Duck, dynamic, strong typing; gradual (since 3.5, but ignored in CPython)
First appeared: February 1991; 30 years ago
Paradigm: Multi-paradigm: object-oriented, procedural (imperative), functional, structured, reflective


In [12]:
s1 = ""

for char in s:
    if not char.isalnum():
        s1 += '**'
    else:
        s1 += char
        
print(s1)

Python**is**an**interpreted**high**level**general**purpose**programming**language****Python**s**design**philosophy**emphasizes**code**readability**with**its**notable**use**of**significant**indentation****Wikipedia**Developer****Python**Software**Foundation**Stable**release****3**9**5******3**May**2021****29**days**ago**Preview**release****3**10**0b1******3**May**2021****29**days**ago**Typing**discipline****Duck****dynamic****strong**typing****gradual****since**3**5****but**ignored**in**CPython****First**appeared****February**1991****30**years**ago**Paradigm****Multi**paradigm****object**oriented****procedural****imperative******functional****structured****reflective


In [14]:
s1 = ""

for char in s:
    if not char.isalnum() and char!= '' and not char.isspace():
        s1 += '**'
    else:
        s1 += char
        
print(s1)

Python is an interpreted high**level general**purpose programming language** Python**s design philosophy emphasizes code readability with its notable use of significant indentation** Wikipedia
Developer** Python Software Foundation
Stable release** 3**9**5 ** 3 May 2021** 29 days ago
Preview release** 3**10**0b1 ** 3 May 2021** 29 days ago
Typing discipline** Duck** dynamic** strong typing** gradual **since 3**5** but ignored in CPython**
First appeared** February 1991** 30 years ago
Paradigm** Multi**paradigm** object**oriented** procedural **imperative**** functional** structured** reflective


In [16]:
ss = s.split()
li = []

for char in ss:
    if char[0].islower():
        li.append(char)
print(li)

['is', 'an', 'interpreted', 'high-level', 'general-purpose', 'programming', 'language.', 'design', 'philosophy', 'emphasizes', 'code', 'readability', 'with', 'its', 'notable', 'use', 'of', 'significant', 'indentation.', 'release:', 'days', 'ago', 'release:', 'days', 'ago', 'discipline:', 'dynamic,', 'strong', 'typing;', 'gradual', 'but', 'ignored', 'in', 'appeared:', 'years', 'ago', 'object-oriented,', 'procedural', 'functional,', 'structured,', 'reflective']


### Regular Expression

a group of character/special characters for matching the pattern

In [17]:
import re

In [18]:
print(re.__doc__)

Support for regular expressions (RE).

This module provides regular expression matching operations similar to
those found in Perl.  It supports both 8-bit and Unicode strings; both
the pattern and the strings being processed can contain null bytes and
characters outside the US ASCII range.

Regular expressions can contain both special and ordinary characters.
Most ordinary characters, like "A", "a", or "0", are the simplest
regular expressions; they simply match themselves.  You can
concatenate ordinary characters, so last matches the string 'last'.

The special characters are:
    "."      Matches any character except a newline.
    "^"      Matches the start of the string.
    "$"      Matches the end of the string or just before the newline at
             the end of the string.
    "*"      Matches 0 or more (greedy) repetitions of the preceding RE.
             Greedy means that it will match as many repetitions as possible.
    "+"      Matches 1 or more (greedy) repetitions of t

### Syntax


- method(pattern,string)

In [19]:
print(re.findall('P',s))

['P', 'P', 'P', 'P', 'P', 'P']


In [20]:
print(re.findall('P..',s))

['Pyt', 'Pyt', 'Pyt', 'Pre', 'Pyt', 'Par']


In [21]:
print(re.findall('a..',s))

['an ', 'al-', 'amm', 'ang', 'age', 'asi', 'ada', 'abl', 'ant', 'ati', 'are', 'ati', 'abl', 'ase', 'ay ', 'ays', 'ago', 'ase', 'ay ', 'ays', 'ago', 'ami', 'adu', 'al ', 'app', 'are', 'ary', 'ars', 'ago', 'ara', 'ara', 'al ', 'ati', 'al,']


In [22]:
print(s)

Python is an interpreted high-level general-purpose programming language. Python's design philosophy emphasizes code readability with its notable use of significant indentation. Wikipedia
Developer: Python Software Foundation
Stable release: 3.9.5 / 3 May 2021; 29 days ago
Preview release: 3.10.0b1 / 3 May 2021; 29 days ago
Typing discipline: Duck, dynamic, strong typing; gradual (since 3.5, but ignored in CPython)
First appeared: February 1991; 30 years ago
Paradigm: Multi-paradigm: object-oriented, procedural (imperative), functional, structured, reflective


In [23]:
print(re.findall('^a..',s))

[]


In [24]:
print(re.findall('^P..',s))

['Pyt']


In [25]:
print(re.findall('\A..',s))

['Py']


In [28]:
print(re.findall('..e$',s))

['ive']


In [29]:
print(re.findall('....e$',s))

['ctive']


In [30]:
s

"Python is an interpreted high-level general-purpose programming language. Python's design philosophy emphasizes code readability with its notable use of significant indentation. Wikipedia\nDeveloper: Python Software Foundation\nStable release: 3.9.5 / 3 May 2021; 29 days ago\nPreview release: 3.10.0b1 / 3 May 2021; 29 days ago\nTyping discipline: Duck, dynamic, strong typing; gradual (since 3.5, but ignored in CPython)\nFirst appeared: February 1991; 30 years ago\nParadigm: Multi-paradigm: object-oriented, procedural (imperative), functional, structured, reflective"

In [32]:
st='jfjjktfkjhiytdm'

print(re.findall('....\Z',st))

['ytdm']


In [33]:
print(re.findall('...',s))

['Pyt', 'hon', ' is', ' an', ' in', 'ter', 'pre', 'ted', ' hi', 'gh-', 'lev', 'el ', 'gen', 'era', 'l-p', 'urp', 'ose', ' pr', 'ogr', 'amm', 'ing', ' la', 'ngu', 'age', '. P', 'yth', "on'", 's d', 'esi', 'gn ', 'phi', 'los', 'oph', 'y e', 'mph', 'asi', 'zes', ' co', 'de ', 'rea', 'dab', 'ili', 'ty ', 'wit', 'h i', 'ts ', 'not', 'abl', 'e u', 'se ', 'of ', 'sig', 'nif', 'ica', 'nt ', 'ind', 'ent', 'ati', 'on.', ' Wi', 'kip', 'edi', 'Dev', 'elo', 'per', ': P', 'yth', 'on ', 'Sof', 'twa', 're ', 'Fou', 'nda', 'tio', 'Sta', 'ble', ' re', 'lea', 'se:', ' 3.', '9.5', ' / ', '3 M', 'ay ', '202', '1; ', '29 ', 'day', 's a', 'Pre', 'vie', 'w r', 'ele', 'ase', ': 3', '.10', '.0b', '1 /', ' 3 ', 'May', ' 20', '21;', ' 29', ' da', 'ys ', 'ago', 'Typ', 'ing', ' di', 'sci', 'pli', 'ne:', ' Du', 'ck,', ' dy', 'nam', 'ic,', ' st', 'ron', 'g t', 'ypi', 'ng;', ' gr', 'adu', 'al ', '(si', 'nce', ' 3.', '5, ', 'but', ' ig', 'nor', 'ed ', 'in ', 'CPy', 'tho', 'Fir', 'st ', 'app', 'ear', 'ed:', ' Fe', 'bru'

In [34]:
print(re.findall('...[a-z]',s))

['Pyth', 'on i', 's an', ' int', 'erpr', 'eted', ' hig', 'h-le', 'el g', 'ener', 'al-p', 'urpo', 'se p', 'rogr', 'ammi', 'ng l', 'angu', '. Py', 'thon', "'s d", 'esig', 'n ph', 'ilos', 'ophy', ' emp', 'hasi', 'es c', 'de r', 'eada', 'bili', 'ty w', 'th i', 'ts n', 'otab', 'le u', 'se o', 'f si', 'gnif', 'ican', 't in', 'dent', 'atio', '. Wi', 'kipe', 'Deve', 'lope', ': Py', 'thon', ' Sof', 'twar', 'e Fo', 'unda', 'tion', 'Stab', 'le r', 'elea', '3 Ma', '29 d', 'ys a', 'Prev', 'ew r', 'elea', '0.0b', '3 Ma', '29 d', 'ys a', 'Typi', 'ng d', 'isci', 'plin', ': Du', 'k, d', 'ynam', 'c, s', 'tron', 'g ty', 'ping', '; gr', 'adua', 'l (s', 'ince', '5, b', 'ut i', 'gnor', 'ed i', ' CPy', 'thon', 'Firs', 't ap', 'pear', ': Fe', 'brua', '30 y', 'ears', ' ago', 'Para', 'digm', ': Mu', 'ti-p', 'arad', 'm: o', 'bjec', 't-or', 'ient', 'd, p', 'roce', 'dura', 'l (i', 'mper', 'ativ', '), f', 'unct', 'iona', 'l, s', 'truc', 'ture', 'd, r', 'efle', 'ctiv']


In [35]:
print(re.findall('[a-z].[a-z]',s))

['yth', 'n i', 's a', 'n i', 'nte', 'rpr', 'ete', 'd h', 'igh', 'lev', 'l g', 'ene', 'ral', 'pur', 'pos', 'e p', 'rog', 'ram', 'min', 'g l', 'ang', 'uag', 'yth', "n's", 'des', 'ign', 'phi', 'los', 'oph', 'y e', 'mph', 'asi', 'zes', 'cod', 'e r', 'ead', 'abi', 'lit', 'y w', 'ith', 'its', 'not', 'abl', 'e u', 'e o', 'f s', 'ign', 'ifi', 'can', 't i', 'nde', 'nta', 'tio', 'iki', 'ped', 'eve', 'lop', 'yth', 'oft', 'war', 'oun', 'dat', 'ion', 'tab', 'e r', 'ele', 'ase', 'day', 's a', 'rev', 'iew', 'rel', 'eas', 'day', 's a', 'ypi', 'g d', 'isc', 'ipl', 'ine', 'uck', 'dyn', 'ami', 'str', 'ong', 'typ', 'ing', 'gra', 'dua', 'sin', 'but', 'ign', 'ore', 'd i', 'yth', 'irs', 't a', 'ppe', 'are', 'ebr', 'uar', 'yea', 's a', 'ara', 'dig', 'ult', 'i-p', 'ara', 'dig', 'obj', 'ect', 'ori', 'ent', 'pro', 'ced', 'ura', 'imp', 'era', 'tiv', 'fun', 'cti', 'ona', 'str', 'uct', 'ure', 'ref', 'lec', 'tiv']


In [37]:
print(re.findall('[A-Za-z].[a-z]',s))

['Pyt', 'hon', 's a', 'n i', 'nte', 'rpr', 'ete', 'd h', 'igh', 'lev', 'l g', 'ene', 'ral', 'pur', 'pos', 'e p', 'rog', 'ram', 'min', 'g l', 'ang', 'uag', 'Pyt', 'hon', 's d', 'esi', 'n p', 'hil', 'oso', 'phy', 'emp', 'has', 'ize', 's c', 'ode', 'rea', 'dab', 'ili', 'y w', 'ith', 'its', 'not', 'abl', 'e u', 'e o', 'f s', 'ign', 'ifi', 'can', 't i', 'nde', 'nta', 'tio', 'Wik', 'ipe', 'dia', 'Dev', 'elo', 'per', 'Pyt', 'hon', 'Sof', 'twa', 'Fou', 'nda', 'tio', 'Sta', 'ble', 'rel', 'eas', 'May', 'day', 's a', 'Pre', 'vie', 'w r', 'ele', 'ase', 'May', 'day', 's a', 'Typ', 'ing', 'dis', 'cip', 'lin', 'Duc', 'dyn', 'ami', 'str', 'ong', 'typ', 'ing', 'gra', 'dua', 'sin', 'but', 'ign', 'ore', 'd i', 'CPy', 'tho', 'Fir', 't a', 'ppe', 'are', 'Feb', 'rua', 'yea', 's a', 'Par', 'adi', 'Mul', 'i-p', 'ara', 'dig', 'obj', 'ect', 'ori', 'ent', 'pro', 'ced', 'ura', 'imp', 'era', 'tiv', 'fun', 'cti', 'ona', 'str', 'uct', 'ure', 'ref', 'lec', 'tiv']


In [38]:
print(re.findall('[A-Za-z][a-z][a-z]',s))

['Pyt', 'hon', 'int', 'erp', 'ret', 'hig', 'lev', 'gen', 'era', 'pur', 'pos', 'pro', 'gra', 'mmi', 'lan', 'gua', 'Pyt', 'hon', 'des', 'ign', 'phi', 'los', 'oph', 'emp', 'has', 'ize', 'cod', 'rea', 'dab', 'ili', 'wit', 'its', 'not', 'abl', 'use', 'sig', 'nif', 'ica', 'ind', 'ent', 'ati', 'Wik', 'ipe', 'dia', 'Dev', 'elo', 'per', 'Pyt', 'hon', 'Sof', 'twa', 'Fou', 'nda', 'tio', 'Sta', 'ble', 'rel', 'eas', 'May', 'day', 'ago', 'Pre', 'vie', 'rel', 'eas', 'May', 'day', 'ago', 'Typ', 'ing', 'dis', 'cip', 'lin', 'Duc', 'dyn', 'ami', 'str', 'ong', 'typ', 'ing', 'gra', 'dua', 'sin', 'but', 'ign', 'ore', 'Pyt', 'hon', 'Fir', 'app', 'ear', 'Feb', 'rua', 'yea', 'ago', 'Par', 'adi', 'Mul', 'par', 'adi', 'obj', 'ect', 'ori', 'ent', 'pro', 'ced', 'ura', 'imp', 'era', 'tiv', 'fun', 'cti', 'ona', 'str', 'uct', 'ure', 'ref', 'lec', 'tiv']


In [39]:
print(re.findall('[A-Z][a-z][a-z]',s))

['Pyt', 'Pyt', 'Wik', 'Dev', 'Pyt', 'Sof', 'Fou', 'Sta', 'May', 'Pre', 'May', 'Typ', 'Duc', 'Pyt', 'Fir', 'Feb', 'Par', 'Mul']


In [40]:
d = 'Today I received 10$ from my Boss'

In [41]:
print(re.findall('..$',d))

['ss']


In [43]:
print(re.findall('\.',s))

['.', '.', '.', '.', '.', '.', '.']


In [44]:
print(re.findall('\W',s))

[' ', ' ', ' ', ' ', '-', ' ', '-', ' ', ' ', '.', ' ', "'", ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', '.', ' ', '\n', ':', ' ', ' ', ' ', '\n', ' ', ':', ' ', '.', '.', ' ', '/', ' ', ' ', ' ', ';', ' ', ' ', ' ', '\n', ' ', ':', ' ', '.', '.', ' ', '/', ' ', ' ', ' ', ';', ' ', ' ', ' ', '\n', ' ', ':', ' ', ',', ' ', ',', ' ', ' ', ';', ' ', ' ', '(', ' ', '.', ',', ' ', ' ', ' ', ' ', ')', '\n', ' ', ':', ' ', ' ', ';', ' ', ' ', ' ', '\n', ':', ' ', '-', ':', ' ', '-', ',', ' ', ' ', '(', ')', ',', ' ', ',', ' ', ',', ' ']


In [45]:
print(re.findall('\d',s))

['3', '9', '5', '3', '2', '0', '2', '1', '2', '9', '3', '1', '0', '0', '1', '3', '2', '0', '2', '1', '2', '9', '3', '5', '1', '9', '9', '1', '3', '0']


In [46]:
print(re.findall('\d.',s))

['3.', '9.', '5 ', '3 ', '20', '21', '29', '3.', '10', '0b', '1 ', '3 ', '20', '21', '29', '3.', '5,', '19', '91', '30']


In [47]:
print(re.findall('\d\d',s))

['20', '21', '29', '10', '20', '21', '29', '19', '91', '30']


In [48]:
print(re.findall('\d\d\d',s))

['202', '202', '199']


In [49]:
print(re.findall('\d\d\d\d',s))

['2021', '2021', '1991']


In [50]:
print(re.findall('\d\d\d\d\d',s))

[]


In [52]:
n = '4556'
n2 = '365'

print(re.findall('^d\d\d$',n))

[]


In [53]:
print(re.findall('^d\d\d$',n2))

[]


In [54]:
print(re.findall('^d\d\d$','3o5'))

[]


In [56]:
print(re.findall('\s',s)

SyntaxError: unexpected EOF while parsing (<ipython-input-56-3e08a03192ff>, line 1)

In [57]:
print(re.findall('\S',s))

['P', 'y', 't', 'h', 'o', 'n', 'i', 's', 'a', 'n', 'i', 'n', 't', 'e', 'r', 'p', 'r', 'e', 't', 'e', 'd', 'h', 'i', 'g', 'h', '-', 'l', 'e', 'v', 'e', 'l', 'g', 'e', 'n', 'e', 'r', 'a', 'l', '-', 'p', 'u', 'r', 'p', 'o', 's', 'e', 'p', 'r', 'o', 'g', 'r', 'a', 'm', 'm', 'i', 'n', 'g', 'l', 'a', 'n', 'g', 'u', 'a', 'g', 'e', '.', 'P', 'y', 't', 'h', 'o', 'n', "'", 's', 'd', 'e', 's', 'i', 'g', 'n', 'p', 'h', 'i', 'l', 'o', 's', 'o', 'p', 'h', 'y', 'e', 'm', 'p', 'h', 'a', 's', 'i', 'z', 'e', 's', 'c', 'o', 'd', 'e', 'r', 'e', 'a', 'd', 'a', 'b', 'i', 'l', 'i', 't', 'y', 'w', 'i', 't', 'h', 'i', 't', 's', 'n', 'o', 't', 'a', 'b', 'l', 'e', 'u', 's', 'e', 'o', 'f', 's', 'i', 'g', 'n', 'i', 'f', 'i', 'c', 'a', 'n', 't', 'i', 'n', 'd', 'e', 'n', 't', 'a', 't', 'i', 'o', 'n', '.', 'W', 'i', 'k', 'i', 'p', 'e', 'd', 'i', 'a', 'D', 'e', 'v', 'e', 'l', 'o', 'p', 'e', 'r', ':', 'P', 'y', 't', 'h', 'o', 'n', 'S', 'o', 'f', 't', 'w', 'a', 'r', 'e', 'F', 'o', 'u', 'n', 'd', 'a', 't', 'i', 'o', 'n',

In [58]:
d

'Today I received 10$ from my Boss'

In [60]:
print(re.findall('\b.','hello'))

[]


In [61]:
print(re.findall('P*',s))

['P', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', 'P', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', 'P', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', 

In [62]:
print(re.findall('P\d*',s))

['P', 'P', 'P', 'P', 'P', 'P']


In [65]:
email = 'manasa_nunna@gmail.com'

In [66]:
print(re.findall('[a-z]*',email))

['manasa', '', 'nunna', '', 'gmail', '', 'com', '']


In [68]:
print(re.findall('[a-z_]*',email))

['manasa_nunna', '', 'gmail', '', 'com', '']


In [69]:
print(re.findall('[a-z]+',email))

['manasa', 'nunna', 'gmail', 'com']


In [70]:
print(re.findall('[a-z]?',email))

['m', 'a', 'n', 'a', 's', 'a', '', 'n', 'u', 'n', 'n', 'a', '', 'g', 'm', 'a', 'i', 'l', '', 'c', 'o', 'm', '']


In [71]:
print(re.findall('[a-z]?.',email))

['ma', 'na', 'sa', '_', 'nu', 'nn', 'a@', 'gm', 'ai', 'l.', 'co', 'm']


In [72]:
print(re.findall('[a-z]?[a-z]',email))

['ma', 'na', 'sa', 'nu', 'nn', 'a', 'gm', 'ai', 'l', 'co', 'm']


In [74]:
print(re.findall('[a-z]{2, 5}',email))

[]


In [79]:
print(re.findall('[a-z]{2}','emmacs'))

['em', 'ma', 'cs']


In [80]:
print(re.findall('[a-z]{3}','emmacs'))

['emm', 'acs']


In [81]:
ip = '163.125.255.125'

ip2 = '172.160.250.5'

In [82]:
re.match('163.125.255.[0-2][0-5][0-5]',ip)

<re.Match object; span=(0, 15), match='163.125.255.125'>

In [84]:
print(re.match('163.125.255.[0-2][0-5][0-5]',ip2))

None


In [85]:
obj = re.match('163.125.255.[0-2][0-5][0-5]',ip)

obj.span()

(0, 15)

In [89]:
ip3 = '163.125.255.125opa'

re.match('163.125.255.[0-2][0-5][0-5]',ip3)

<re.Match object; span=(0, 15), match='163.125.255.125'>

In [90]:
ip3 = '163.125.255.125opa'

re.match('163.125.255.[0-2][0-5][0-5]',ip3)

<re.Match object; span=(0, 15), match='163.125.255.125'>

In [91]:
re.match('163.125.255.[0-2][0-5][0-5]',ip)

<re.Match object; span=(0, 15), match='163.125.255.125'>

In [92]:
ip3 = '163.125.255.125opa'

re.match('163.125.255.[0-2][0-5][0-5]$',ip)

<re.Match object; span=(0, 15), match='163.125.255.125'>

In [95]:
ip3 = '163.125.255.125opa'

print(re.match('163.125.255.[0-2][0-5][0-5]$',ip3))

None


In [97]:
re.match('^163.125.255.[0-2][0-5][0-5]$',ip)

<re.Match object; span=(0, 15), match='163.125.255.125'>

In [98]:
obj.string

'163.125.255.125'

- +91
- 6789
- 10

In [103]:
m1 = '9876543210'
m2 = '3216549870'
m3 = '987654321o'

In [104]:

pattern = '^[6789]\d{9}$'

re.match(pattern,m1)

<re.Match object; span=(0, 10), match='9876543210'>

In [105]:
print(re.match(pattern,m2), re.match(pattern,m3))

None None


In [107]:
m4 = '+919876543210'

pattern = '+91^[6789]\d{9}'

In [109]:
re.match(pattern,m4)

error: nothing to repeat at position 0

In [111]:
pattern = '^[+]91[6789]\d{9}'

re.match(pattern,m4)

<re.Match object; span=(0, 13), match='+919876543210'>

In [112]:
m3 = '+9164578d9656'

print(re.match(pattern,m3))

None


### Task

### userName@domain.extension

- userName
    - a-zA-S0-9._
    - it shouldn't startwith
    - min 8 max 15
- @
- Domain
    - a-zA-Z0-9
    - min 2 max 10
- Extension
    - a-zA-Z
    - {2,5}

### sub -> replace

### Synatx

sub(Pat,repla,string)

In [113]:
print(re.sub('\s','-',s))

Python-is-an-interpreted-high-level-general-purpose-programming-language.-Python's-design-philosophy-emphasizes-code-readability-with-its-notable-use-of-significant-indentation.-Wikipedia-Developer:-Python-Software-Foundation-Stable-release:-3.9.5-/-3-May-2021;-29-days-ago-Preview-release:-3.10.0b1-/-3-May-2021;-29-days-ago-Typing-discipline:-Duck,-dynamic,-strong-typing;-gradual-(since-3.5,-but-ignored-in-CPython)-First-appeared:-February-1991;-30-years-ago-Paradigm:-Multi-paradigm:-object-oriented,-procedural-(imperative),-functional,-structured,-reflective


In [114]:
print(re.sub('\d','0',s))

Python is an interpreted high-level general-purpose programming language. Python's design philosophy emphasizes code readability with its notable use of significant indentation. Wikipedia
Developer: Python Software Foundation
Stable release: 0.0.0 / 0 May 0000; 00 days ago
Preview release: 0.00.0b0 / 0 May 0000; 00 days ago
Typing discipline: Duck, dynamic, strong typing; gradual (since 0.0, but ignored in CPython)
First appeared: February 0000; 00 years ago
Paradigm: Multi-paradigm: object-oriented, procedural (imperative), functional, structured, reflective


In [115]:
print(re.sub('\S','*',s))

****** ** ** *********** ********** *************** *********** ********* ******** ****** ********** ********** **** *********** **** *** ******* *** ** *********** ************ *********
********** ****** ******** **********
****** ******** ***** * * *** ***** ** **** ***
******* ******** ******** * * *** ***** ** **** ***
****** *********** ***** ******** ****** ******* ******* ****** **** *** ******* ** ********
***** ********* ******** ***** ** ***** ***
********* *************** **************** ********** ************* *********** *********** **********


In [116]:
print(re.subn('\S','*',s))

('****** ** ** *********** ********** *************** *********** ********* ******** ****** ********** ********** **** *********** **** *** ******* *** ** *********** ************ *********\n********** ****** ******** **********\n****** ******** ***** * * *** ***** ** **** ***\n******* ******** ******** * * *** ***** ** **** ***\n****** *********** ***** ******** ****** ******* ******* ****** **** *** ******* ** ********\n***** ********* ******** ***** ** ***** ***\n********* *************** **************** ********** ************* *********** *********** **********', 492)


In [117]:
print(re.sub('[aeiou]','*',s))

Pyth*n *s *n *nt*rpr*t*d h*gh-l*v*l g*n*r*l-p*rp*s* pr*gr*mm*ng l*ng**g*. Pyth*n's d*s*gn ph*l*s*phy *mph*s*z*s c*d* r**d*b*l*ty w*th *ts n*t*bl* *s* *f s*gn*f*c*nt *nd*nt*t**n. W*k*p*d**
D*v*l*p*r: Pyth*n S*ftw*r* F**nd*t**n
St*bl* r*l**s*: 3.9.5 / 3 M*y 2021; 29 d*ys *g*
Pr*v**w r*l**s*: 3.10.0b1 / 3 M*y 2021; 29 d*ys *g*
Typ*ng d*sc*pl*n*: D*ck, dyn*m*c, str*ng typ*ng; gr*d**l (s*nc* 3.5, b*t *gn*r*d *n CPyth*n)
F*rst *pp**r*d: F*br**ry 1991; 30 y**rs *g*
P*r*d*gm: M*lt*-p*r*d*gm: *bj*ct-*r**nt*d, pr*c*d*r*l (*mp*r*t*v*), f*nct**n*l, str*ct*r*d, r*fl*ct*v*


In [119]:
print(re.subn('[^aeiou]','*',s))

('****o**i**a**i**e***e*e***i****e*e***e*e*a***u**o*e***o**a**i****a**ua*e******o*****e*i*****i*o*o****e***a*i*e***o*e**ea*a*i*i****i***i****o*a**e*u*e*o***i**i*i*a***i**e**a*io****i*i*e*ia**e*e*o*e*******o***o***a*e**ou**a*io****a**e**e*ea*e*************a************a***a*o***e*ie***e*ea*e****************a************a***a*o****i****i**i**i*e***u*******a*i******o******i******a*ua****i**e*******u**i**o*e**i*******o****i****a**ea*e****e**ua*************ea***a*o**a*a*i*****u**i**a*a*i****o**e***o*ie**e*****o*e*u*a***i**e*a*i*e****u***io*a******u**u*e****e**e**i*e', 402)


In [120]:
print(re.subn('[^aeiou\s]','*',s))

('****o* i* a* i**e***e*e* *i****e*e* *e*e*a***u**o*e **o**a**i** *a**ua*e* ****o*** *e*i** **i*o*o*** e***a*i*e* *o*e *ea*a*i*i** *i** i** *o*a**e u*e o* *i**i*i*a** i**e**a*io** *i*i*e*ia\n*e*e*o*e** ****o* *o***a*e *ou**a*io*\n**a**e *e*ea*e* ***** * * *a* ***** ** *a** a*o\n**e*ie* *e*ea*e* ******** * * *a* ***** ** *a** a*o\n***i** *i**i**i*e* *u*** ***a*i** ***o** ***i*** **a*ua* **i**e **** *u* i**o*e* i* *****o**\n*i*** a**ea*e** *e**ua** ***** ** *ea** a*o\n*a*a*i*** *u**i**a*a*i*** o**e***o*ie**e** **o*e*u*a* *i**e*a*i*e** *u***io*a** ***u**u*e** *e**e**i*e', 329)
